src/util-ebpf.c

```diff
@@ -313,141 +497,196 @@ int EBPFSetupXDP(const char *iface, int fd, uint8_t flags)
+                                uint32_t hash, struct timespec *ctime,
+                                uint64_t pkts_cnt, uint64_t bytes_cnt)
 {
-    SCLogDebug("Got curtime %" PRIu64 " and value %" PRIu64 " (sp:%d, dp:%d)",
-               curtime->tv_sec, value->time / 1000000000,
-               key->port16[0], key->port16[1]
-              );
-
-    if (curtime->tv_sec - value->time / 1000000000 > BYPASSED_FLOW_TIMEOUT) {
-        SCLogDebug("Got no packet for %d -> %d at %" PRIu64,
-                   key->port16[0], key->port16[1], value->time);
-        return 1;
+    Flow *f = FlowGetExistingFlowFromHash(flow_key, hash);
+    if (f != NULL) {
+        SCLogDebug("bypassed flow found %d -> %d, doing accounting",
+                    f->sp, f->dp);
+        FlowCounters *fc = FlowGetStorageById(f, GetFlowBypassCounterID());
+        if (fc == NULL) {
+            FLOWLOCK_UNLOCK(f);
+            flowstats->count++;
+            return true;
+        }
+        if (flow_key->sp == f->sp) {
+            if (pkts_cnt != fc->todstpktcnt) {
+                flowstats->packets += pkts_cnt - fc->todstpktcnt;
+                flowstats->bytes += bytes_cnt - fc->todstbytecnt;
+                fc->todstpktcnt = pkts_cnt;
+                fc->todstbytecnt = bytes_cnt;
+                /* interval based so no meaning to update the millisecond.
+                 * Let's keep it fast and simple */
+                f->lastts.tv_sec = ctime->tv_sec;
+            }
+        } else {
+            if (pkts_cnt != fc->tosrcpktcnt) {
+                flowstats->packets += pkts_cnt - fc->tosrcpktcnt;
+                flowstats->bytes += bytes_cnt - fc->tosrcbytecnt;
+                fc->tosrcpktcnt = pkts_cnt;
+                fc->tosrcbytecnt = bytes_cnt;
+                f->lastts.tv_sec = ctime->tv_sec;
+            }
+        }
+        FLOWLOCK_UNLOCK(f);
+        return false;
+    } else {
+        /* Flow has disappeared so it has been removed due to timeout.
+         * This means we did not see any packet since bypassed_timeout/flow_dump_interval
+         * checks so we are highly probably have no packet to account.
+         * So we just discard the flow */
+        SCLogDebug("No flow for %d -> %d", flow_key->sp, flow_key->dp);
+        SCLogDebug("Dead with pkts %lu bytes %lu", pkts_cnt, bytes_cnt);
+        flowstats->count++;
+        return true;
     }
-    return 0;
 }
 
+typedef bool (*OpFlowForKey)(struct flows_stats *flowstats, FlowKey *flow_key,
+                            uint32_t hash, struct timespec *ctime,
+                            uint64_t pkts_cnt, uint64_t bytes_cnt);
+
 /**
  * Bypassed flows cleaning for IPv4
  *
  * This function iterates on all the flows of the IPv4 table
  * looking for timeouted flow to delete from the flow table.
  */
-static int EBPFForEachFlowV4Table(LiveDevice *dev, const char *name,
+static int EBPFForEachFlowV4Table(ThreadVars *th_v, LiveDevice *dev, const char *name,
```

Comment:
I added a print statement to the start and end of this func to get a sense for how long it runs and for how many entries it finds. 3 observations:
```
[2206] 6/6/2019 -- 11:15:37 - (util-ebpf.c:324) <Info> (EBPFLoadFile) -- Loaded pinned maps, will use already loaded eBPF filter
[2206] 6/6/2019 -- 11:15:37 - (runmode-af-packet.c:533) <Info> (ParseAFPConfig) -- Loaded pinned maps from sysfs
[2206] 6/6/2019 -- 11:15:37 - (util-runmodes.c:297) <Info> (RunModeSetLiveCaptureWorkersForDevice) -- Going to use 12 thread(s)
[2242] 6/6/2019 -- 11:15:37 - (util-ebpf.c:616) <Notice> (EBPFForEachFlowV4Table) -- start
[2206] 6/6/2019 -- 11:15:37 - (util-conf.c:115) <Info> (ConfUnixSocketIsEnable) -- Running in live mode, activating unix socket
[2206] 6/6/2019 -- 11:15:37 - (unix-manager.c:131) <Info> (UnixNew) -- Using unix socket file '/var/run/suricata/suricata-command.socket'
[2206] 6/6/2019 -- 11:15:37 - (tm-threads.c:2157) <Notice> (TmThreadWaitOnThreadInit) -- all 12 packet processing threads, 9 management threads initialized, engine started.
[2235] 6/6/2019 -- 11:15:40 - (source-af-packet.c:510) <Info> (AFPPeersListReachedInc) -- All AFP capture threads are running.
[2242] 6/6/2019 -- 11:16:00 - (util-ebpf.c:692) <Notice> (EBPFForEachFlowV4Table) -- end: hash_cnt 260908
[2242] 6/6/2019 -- 11:16:02 - (util-ebpf.c:616) <Notice> (EBPFForEachFlowV4Table) -- start
[2242] 6/6/2019 -- 11:16:23 - (util-ebpf.c:692) <Notice> (EBPFForEachFlowV4Table) -- end: hash_cnt 262144
[2242] 6/6/2019 -- 11:16:35 - (util-ebpf.c:616) <Notice> (EBPFForEachFlowV4Table) -- start
[2242] 6/6/2019 -- 11:16:56 - (util-ebpf.c:692) <Notice> (EBPFForEachFlowV4Table) -- end: hash_cnt 262144
[2242] 6/6/2019 -- 11:17:07 - (util-ebpf.c:616) <Notice> (EBPFForEachFlowV4Table) -- start
[2242] 6/6/2019 -- 11:17:35 - (util-ebpf.c:692) <Notice> (EBPFForEachFlowV4Table) -- end: hash_cnt 287759
[2242] 6/6/2019 -- 11:17:46 - (util-ebpf.c:616) <Notice> (EBPFForEachFlowV4Table) -- start
[2242] 6/6/2019 -- 11:18:22 - (util-ebpf.c:692) <Notice> (EBPFForEachFlowV4Table) -- end: hash_cnt 359593
[2242] 6/6/2019 -- 11:18:33 - (util-ebpf.c:616) <Notice> (EBPFForEachFlowV4Table) -- start
[2242] 6/6/2019 -- 11:18:57 - (util-ebpf.c:692) <Notice> (EBPFForEachFlowV4Table) -- end: hash_cnt 274873
[2242] 6/6/2019 -- 11:19:09 - (util-ebpf.c:616) <Notice> (EBPFForEachFlowV4Table) -- start
[2242] 6/6/2019 -- 11:19:33 - (util-ebpf.c:692) <Notice> (EBPFForEachFlowV4Table) -- end: hash_cnt 273435
[2242] 6/6/2019 -- 11:19:44 - (util-ebpf.c:616) <Notice> (EBPFForEachFlowV4Table) -- start
[2242] 6/6/2019 -- 11:20:16 - (util-ebpf.c:692) <Notice> (EBPFForEachFlowV4Table) -- end: hash_cnt 334341
[2242] 6/6/2019 -- 11:20:27 - (util-ebpf.c:616) <Notice> (EBPFForEachFlowV4Table) -- start
[2242] 6/6/2019 -- 11:20:55 - (util-ebpf.c:692) <Notice> (EBPFForEachFlowV4Table) -- end: hash_cnt 300721
```

1. it starts before af-packet is ready. Not sure if this is intentional.
2. it seems to find more entries than my .max_entries setting of 262144
3. runtime is really long 20 seconds is no exception.

So it seems for that ~20 seconds we're doing 262k+ flow table lookups and updates. That seems to be an expensive operation.

Reply:
Ok the more than 256k thing seems to be explained by the fact that while we loop and remove dead flows, new ones get inserted and we loop over (part of) those as well:
```
[13510] 6/6/2019 -- 11:30:05 - (util-ebpf.c:616) <Notice> (EBPFForEachFlowV4Table) -- start
[13510] 6/6/2019 -- 11:30:26 - (util-ebpf.c:693) <Notice> (EBPFForEachFlowV4Table) -- end: hash_cnt 262144, dead 0, diff 262144
[13510] 6/6/2019 -- 11:30:38 - (util-ebpf.c:616) <Notice> (EBPFForEachFlowV4Table) -- start
[13510] 6/6/2019 -- 11:31:09 - (util-ebpf.c:693) <Notice> (EBPFForEachFlowV4Table) -- end: hash_cnt 306348, dead 125551, diff 180797
[13510] 6/6/2019 -- 11:31:20 - (util-ebpf.c:616) <Notice> (EBPFForEachFlowV4Table) -- start
[13510] 6/6/2019 -- 11:31:55 - (util-ebpf.c:693) <Notice> (EBPFForEachFlowV4Table) -- end: hash_cnt 357770, dead 132159, diff 225611
```

Reply:
In my test it seems about 1/4 to 1/3 of the flows are 'dead' each time. I wonder if it would make sense to skip the newly added ones? I guess this would require adding a field to 'struct pair'. I think the newly added flows can't time out yet because of FLOW_BYPASSED_TIMEOUT == 100 seconds?

Reply:
I guess we still want to inspect them for their stats though.

Reply:
We dump the flow table from the kernel/userspace map so we can't skip some. And also we get the stats.